File: sprintboot-Security&JWT/jwt/src/main/java/com/cos/jwt/config/SecurityConfig.java
package com.cos.jwt.config;

import com.cos.jwt.config.jwt.JwtAuthenticationFilter;
import com.cos.jwt.filter.Myfilter1;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;
import org.springframework.web.filter.CorsFilter;

@Configuration
@EnableWebSecurity
@RequiredArgsConstructor
public class SecurityConfig {

    private final CorsFilter corsFilter;
    private AuthenticationManager authenticationManager;

    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {

        http
                .httpBasic(AbstractHttpConfigurer::disable)
                .csrf(AbstractHttpConfigurer::disable)
                .formLogin(AbstractHttpConfigurer::disable)
//                .addFilterBefore(new Myfilter1(), BasicAuthenticationFilter.class)
                .addFilter(new JwtAuthenticationFilter(authenticationManager))//AuthenticationManager
                .sessionManagement(session ->
                        session.sessionCreationPolicy(SessionCreationPolicy.STATELESS)); // 세션을 사용하징 않는다. 1. 서버가 여러개가있다면 서버마다 세션 영역 메모리를 따로 만들어야함 2. 서버에서 http-only는 서버에서 기본적으로 요청을 거부하지만 false걸어버리면 요청이 오지만 보안적으로 안좋음

        http
                .authorizeHttpRequests((authorizeRequests) ->
                        authorizeRequests
                                .requestMatchers("/api/v1/user/**").hasAnyRole("USER", "MANAGER", "ADMIN") // 인증만 되면 들어갈수 있는 주소
                                .requestMatchers("/api/v1/manager/**").hasAnyRole("MANAGER", "ADMIN")
                                .requestMatchers("/api/v1/admin/**").hasAnyRole("ADMIN") // 인증만 되면 들어갈수 있는 주소
                                .anyRequest().permitAll()

                );
        return http.build();
    }


}

File: kotlin-spring-security/demo/src/main/kotlin/com/example/demo/common/authority/SecurityConfig.kt
package com.example.demo.common.authority

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.factory.PasswordEncoderFactories
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter

@Configuration
@EnableWebSecurity
class SecurityConfig(
    private val jwtTokenProvider: JwtTokenProvider
) {
    // SecurityFilterChain을 정의하는 빈 메서드
    @Bean
    fun filterChain(http: HttpSecurity): SecurityFilterChain {
        http
            // HTTP Basic 인증 비활성화
            .httpBasic { it.disable() }
            // CSRF(Cross-Site Request Forgery) 비활성화
            .csrf { it.disable() }
            // 세션 관리 정책을 STATELESS로 설정하여 세션을 사용하지 않음
            .sessionManagement {
                it.sessionCreationPolicy(SessionCreationPolicy.STATELESS)
            }
            // 특정 API 엔드포인트에 대한 접근 권한 설정
            .authorizeHttpRequests {
                it.requestMatchers("/api/member/signup", "/api/member/login").anonymous()
                    .requestMatchers("/api/member/info/**").hasRole("MEMBER")
                    .anyRequest().permitAll() // 나머지 요청은 모두 허용
            }
            // JwtAuthenticationFilter를 UsernamePasswordAuthenticationFilter 앞에 추가
            .addFilterBefore(
                JwtAuthenticationFilter(jwtTokenProvider),
                UsernamePasswordAuthenticationFilter::class.java
            )

        // SecurityFilterChain을 빌드하여 반환
        return http.build()
    }

    // PasswordEncoder 빈 등록
    @Bean
    fun passwordEncoder(): PasswordEncoder =
        PasswordEncoderFactories.createDelegatingPasswordEncoder()
}
